src/gateways/googleAnalyticsClient.ts

```diff
@@ -0,0 +1,78 @@
+import { assertWindow } from '../app/utils';
+
+interface Command {
+  name: string;
+  params: Array<any>;
+}
+
+class GoogleAnalyticsClient {
+
+  private trackerNames: Array<string> = [];
+  private pendingCommands: Array<Command> = [];
+
+  private GA = assertWindow().ga;
+
+  private saveCommandForLater(command: Command) {
+    this.pendingCommands.push(command);
+  }
+
+  private flushPendingCommands() {
+    for (let pendingCommand of this.pendingCommands) {
+      this.commandEachTracker(pendingCommand);
+    }
+    this.pendingCommands = [];
+  }
+
+  private commandEachTracker(command: Command) {
+    for (let trackerName of this.trackerNames) {
+      this.GA(trackerName + '.' + command.name, ...command.params);
+    }
+  }
+
+  private isReadyForCommands() {
+    return (this.trackerNames.length > 0);
+  }
+
+  public ga(command_name: string, ...params: any[]) {
+    let command = {name: command_name, params: params} as Command;
+
+    // The client may not be fully initialized by the time the first
+    // command comes in; in that case, save the command until that
+    // initialization.
+
+    if (this.isReadyForCommands()) {
+      this.flushPendingCommands();
+      this.commandEachTracker(command);
+    }
+    else {
+      this.saveCommandForLater(command);
+    }
+  };
+
```

Comment:
I'd suggest calling this something other than `ga`, perhaps `record` or `track`?

Rather than call `this.isReadyForCommands()` on every call, how about getting clever and have the something like:

```javascript
class GoogleAnalyticsClient {
  // by default 'record' simply records the command for later recording
  public record(cmd) { this.pending.push(cmd); } 
  // this method will record to the real GA
  private recordReal(cmd) { ... }

 setTrackingIds(...) {
    flush();
   // replace the record method with the real tracking 
   this.record = this.recordReal;
 } 
}
```

 I have no idea if typescript will agree with this, it will probably require some casting

Reply:
Good idea on flushing in `setTrackingIds`.  

I kind of like exposing `ga` so that clients can know they can call whatever they would have called on the normal `ga` function.  Know what I mean?

Reply:
not sure about swapping out the functions, but you could also have the `ga` function always queue the command and then always call flush, then have flush noop if the client isn't ready